core/src/test/scala/MonitoringSpec.scala

```diff
@@ -553,4 +555,168 @@ object MonitoringSpec extends Properties("monitoring") {
     go || go || go
   }
 
+  /*
+  property("keys are being removed upon end of data from host") = secure {
+    val M = Monitoring.default
+    val i = new Instruments(monitoring = M)
+    val gauge = i.gauge("someGauge",0.0)
+
+    M.keySenescence(Events.every(1 milliseconds),Process.eval(Task.delay(gauge.key))).run.run
+    //TODO: Must refactor to allow simulation of stopped data
+    M.keys.get.run.contains(gauge.key)
+  }
+  */
+
+  property("keys are being removed upon disconnect") = secure {
+    val M = Monitoring.default
+    val i = new Instruments(monitoring = M)
+    val gauge = i.gauge("someGauge",0.0)
+    val key = gauge.key
+
+    //closing the signal, simulating the host dropping the connection
+    M.get(key).close.run
+
+    val x = M.keySenescence(Events.every(1 milliseconds),Process.eval(Task.delay(gauge.key)))
+
+    Thread.sleep(1000)
+
+    x.run.run
+    !M.keys.get.run.contains(key)
+  }
+
+
+  property("TCP disconnect removes host from /mirror/sources list") = secure {
+    val M = Monitoring.default
+    val enqueueSink: Sink[Task, Command] = M.mirroringQueue.enqueue
+    val uri= URI.create("http://localhost")
+    val clusterName = "clusterName"
+    val mirror = Mirror(uri,clusterName)
+    val discard = Discard(uri)
+    val datapoint = new Datapoint[Any](Key[String]("key",Units.TrafficLight),"green")
+    val commands: Process[Task, Command] = Process.emitAll(Seq(mirror, discard))
+
+    //this should be commands.to(enqueueSink) but scala 2.10 didn't like it
+    //val commandEnqueue = commands.to(enqueueSink)
+    val commandEnqueue = commands.zipWith(enqueueSink)((o,f) => f(o)).eval
+
+    val mockParse: URI => Process[Task, Datapoint[Any]] = 
+      _ => Process.eval(Task.delay(throw new RuntimeException("boom")))
+
+    //enqueue the commands
+    commandEnqueue.run.run
+
+    //This is used as a flag to cancel the processing
+    val b = new java.util.concurrent.atomic.AtomicBoolean(false)
+
+    //start processing commands
+    M.processMirroringEvents(
```

Comment:
Not sure if im missing something here, but could you not do:

```
//instead of .runAsyncInterruptibly(_ => (), b)
).timed(1.seconds.toMillis).run
```

Timed delegates to the underpinning of `runAsyncInterruptibly` and should give you the same semantics.